### ISSUE TITLE
Enabling NetworkPolicy on a Regional cluster causes issues

### ISSUE THREAD

**Issue by Stono** [terraform-providers/terraform-provider-google#1451]
Sorry for the broad title, but I didn't know how to word this best.

Long story short if you do this:

  - build a cluster with network policy enabled
  - add a new node pool to that cluster
  - run a manual command to delete the default node pool (as we don't want it)

#3 will fail:

```
module.kubernetes.google_container_node_pool.custom-pool: Provisioning with 'local-exec'...
module.kubernetes.google_container_node_pool.custom-pool (local-exec): Executing: ["/bin/sh" "-c" "gcloud beta --quiet container node-pools delete default-pool --region europe-west4 --cluster delivery-platform --project at-delivery-platform-phoenix"]
module.kubernetes.google_container_node_pool.custom-pool (local-exec): ERROR: (gcloud.beta.container.node-pools.delete) ResponseError: code=400, message=Operation operation-1525805696523-5007b8e4 is currently upgrading cluster delivery-platform. Please wait and try again once it's done.
```

I found it interesting when I queried the pending operations:

```
❯ gcloud beta container operations list --region europe-west4 --project at-delivery-platform-phoenix
NAME                              TYPE              LOCATION      TARGET             STATUS_MESSAGE  STATUS   START_TIME                      END_TIME
operation-1525802640812-401a1b1e  CREATE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T18:04:00.812152363Z  2018-05-08T18:06:15.412706238Z
operation-1525802782925-f870e98e  CREATE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T18:06:22.925390954Z  2018-05-08T18:07:31.867683423Z
operation-1525802853891-ee25dba2  DELETE_NODE_POOL  europe-west4  default-pool                       DONE     2018-05-08T18:07:33.891123078Z  2018-05-08T18:09:07.629271151Z
operation-1525803046703-287142d8  DELETE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T18:10:46.703054094Z  2018-05-08T18:12:21.128558765Z
operation-1525803141575-0b854260  DELETE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T18:12:21.575070852Z  2018-05-08T18:14:24.534328274Z
operation-1525805435726-5fb42ea1  CREATE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T18:50:35.726409556Z  2018-05-08T18:53:13.03940448Z
operation-1525805597864-93e002fa  CREATE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T18:53:17.864017082Z  2018-05-08T18:54:50.580741111Z
operation-1525805696523-5007b8e4  UPGRADE_MASTER    europe-west4  delivery-platform                  RUNNING  2018-05-08T18:54:56.523369416Z
```

CREATE_NODE_POOL completed, so the mutex was unlocked, but then it immediately went into an UPGRADE_MASTER state and I have no idea why, the master wasn't being upgraded.

This doesn't happen with networkpolicy disabled.  I don't know if anyone on the google side of the fence could chip in to if there is actually two actions that need to be waited for, before a node pool is considered built?

**Comment by Stono** [comment]
Yay for inconsistent behaviour. 
The first three times I ran this, I saw what I showed you above.

The fourth time (after I added some operational wait logic to the bash script), it hasn't done the UPGRADE_MASTER and it works:

```
NAME                              TYPE              LOCATION      TARGET             STATUS_MESSAGE  STATUS   START_TIME                      END_TIME
operation-1525802640812-401a1b1e  CREATE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T18:04:00.812152363Z  2018-05-08T18:06:15.412706238Z
operation-1525802782925-f870e98e  CREATE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T18:06:22.925390954Z  2018-05-08T18:07:31.867683423Z
operation-1525802853891-ee25dba2  DELETE_NODE_POOL  europe-west4  default-pool                       DONE     2018-05-08T18:07:33.891123078Z  2018-05-08T18:09:07.629271151Z
operation-1525803046703-287142d8  DELETE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T18:10:46.703054094Z  2018-05-08T18:12:21.128558765Z
operation-1525803141575-0b854260  DELETE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T18:12:21.575070852Z  2018-05-08T18:14:24.534328274Z

operation-1525805435726-5fb42ea1  CREATE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T18:50:35.726409556Z  2018-05-08T18:53:13.03940448Z
operation-1525805597864-93e002fa  CREATE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T18:53:17.864017082Z  2018-05-08T18:54:50.580741111Z
operation-1525805696523-5007b8e4  UPGRADE_MASTER    europe-west4  delivery-platform                  DONE     2018-05-08T18:54:56.523369416Z  2018-05-08T19:02:24.13907098Z
operation-1525806217657-ea420cac  DELETE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T19:03:37.657916539Z  2018-05-08T19:04:44.163981747Z
operation-1525806301694-3ee96658  DELETE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T19:05:01.694809092Z  2018-05-08T19:07:53.193653769Z

operation-1525806654366-15a33f92  CREATE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T19:10:54.366724243Z  2018-05-08T19:13:10.801906983Z
operation-1525806796374-36bad721  CREATE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T19:13:16.374553505Z  2018-05-08T19:14:47.496262641Z
operation-1525806897086-dcbc7b81  UPGRADE_MASTER    europe-west4  delivery-platform                  DONE     2018-05-08T19:14:57.086049212Z  2018-05-08T19:21:41.054843942Z
operation-1525807568822-e951c33e  DELETE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T19:26:08.82232444Z   2018-05-08T19:26:53.717581795Z
operation-1525807623544-32a9f9e3  DELETE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T19:27:03.544887922Z  2018-05-08T19:29:06.676922726Z

operation-1525807926979-018f0b3b  CREATE_CLUSTER    europe-west4  delivery-platform                  DONE     2018-05-08T19:32:06.979346997Z  2018-05-08T19:34:21.233231926Z
operation-1525808068955-2c10f8ef  CREATE_NODE_POOL  europe-west4  custom-pool                        DONE     2018-05-08T19:34:28.955098339Z  2018-05-08T19:35:41.985146313Z
operation-1525808153247-00492e31  DELETE_NODE_POOL  europe-west4  default-pool                       RUNNING  2018-05-08T19:35:53.247600239Z
```

**Comment by ndmckinley** [comment]
Thanks for the report.  Can you tell me - does it work if you set `remove_default_node_pool = true` in the config?

**Comment by Stono** [comment]
Hey @ndmckinley - I did not even know about that flag!
It worked, but I saw another strange thing happen, I ended up with a bonus additional node in one of the zones?:

```
❯ k get nodes -o wide
NAME                                              STATUS    ROLES     AGE         VERSION        EXTERNAL-IP      OS-IMAGE                             KERNEL-VERSION   CONTAINER-RUNTIME
gke-delivery-platform-custom-pool-1e67e5b7-sn0r   Ready     <none>    1m          v1.9.7-gke.0   35.204.225.26    Container-Optimized OS from Google   4.4.111+         docker://17.3.2
gke-delivery-platform-custom-pool-3e4f8a7f-6mqs   Ready     <none>    <invalid>   v1.9.7-gke.0   35.204.87.158    Container-Optimized OS from Google   4.4.111+         docker://17.3.2
gke-delivery-platform-custom-pool-3e4f8a7f-ht5f   Ready     <none>    1m          v1.9.7-gke.0   35.204.130.232   Container-Optimized OS from Google   4.4.111+         docker://17.3.2
gke-delivery-platform-custom-pool-482b3d6e-k8td   Ready     <none>    1m          v1.9.7-gke.0   35.204.120.189   Container-Optimized OS from Google   4.4.111+         docker://17.3.2
```

**Comment by ndmckinley** [comment]
Glad it worked!  Unfortunately we don't really interact with the API the way kubectl does - we're a wrapper around the GKE apis.  Does the bonus node show up in [nodePools.get](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.zones.clusters.nodePools/get)?

**Comment by thomasriley** [comment]
Hey! Worth noting that the original issue raised may be the same `UPGRADE_MASTER` issue, in which the K8 masters are upgraded, raised in #1643 that is fixed via #1660 in the next release.

**Comment by paddycarver** [comment]
This looks like a fixed issue, so I'm going to close it out. @Stono, if you are reproducibly getting an extra node, I'd love to see a new issue for it, ideally with [debug logs](https://www.terraform.io/docs/internals/debugging.html) for the apply that created the cluster and node pool--though even then, I don't believe we interact with things at the node level, so I'm not sure what we'll be able to do, but we can at least take a look and escalate if it's an upstream problem.
